### PR TITLE
initial commit

### PR DESCRIPTION
**Pull request by alexjoybc**
# Description

This PR includes the following proposed change(s):

- Add initial code

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactoring / Documentation
- [ ] Version change

if your change is a breaking change, please add `breaking change` label to this PR

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

Nope

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/src/main/java/ca/bc/gov/open/pssg/docmerge/config/ConfigProperties.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
+package ca.bc.gov.open.pssg.docmerge.config;
+
+import org.springframework.boot.context.properties.ConfigurationProperties;
+
+/**
+ * 
+ * Externalize configuration for easy access to properties either as OpenShift Secrets or local Env variables. 
+ * 
+ * @author shaunmillargov
```
Comment (akroon3r) [inline src/src/main/java/ca/bc/gov/open/pssg/docmerge/config/ConfigProperties.java:9]: if this was modified it should be , alexjoy, taygov ....etc

File: src/Still_to_be_done.txt [hunk 2]
```diff
@@ -0,0 +1,3 @@
+- add document bookmark page and flag (done).
```
Comment (TayGov) [inline src/Still_to_be_done.txt:1]: ??

File: src/pom.xml [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+			<scope>test</scope>
+			<exclusions>
+				<exclusion>
+					<groupId>org.junit.vintage</groupId>
+					<artifactId>junit-vintage-engine</artifactId>
+				</exclusion>
+			</exclusions>
+		</dependency>
+
+<!--
```
Comment (TayGov) [inline src/pom.xml:52]: We should drop this

File: src/src/main/java/ca/bc/gov/open/pssg/docmerge/config/ConfigProperties.java (hunk 1 above)
Comment (alexjoybc) [reply src/src/main/java/ca/bc/gov/open/pssg/docmerge/config/ConfigProperties.java:9]: oh no

File: src/src/main/java/ca/bc/gov/open/pssg/docmerge/model/DocMergeResponse.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+	public String getMimeType() {
+		return mimeType;
+	}
+
+	@JsonProperty("mimeType")
+	public void setMimeType(String mimeType) {
+		this.mimeType = mimeType;
+	}
+
+}
```
Comment (TayGov) [inline src/src/main/java/ca/bc/gov/open/pssg/docmerge/model/DocMergeResponse.java:36]: new line

File: src/src/main/java/ca/bc/gov/open/pssg/docmerge/config/ConfigProperties.java (hunk 1 above)
Comment (alexjoybc) [reply src/src/main/java/ca/bc/gov/open/pssg/docmerge/config/ConfigProperties.java:9]: after first merge

File: src/src/main/java/ca/bc/gov/open/pssg/docmerge/controller/MergeController.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+	private final Logger logger = LoggerFactory.getLogger(MergeController.class);
+
+	@PostMapping(value = {"/merge/{correlationId}" }, 
+			consumes = DocMergeConstants.JSON_CONTENT, 
+			produces = DocMergeConstants.JSON_CONTENT)
+	public ResponseEntity<JSONResponse<DocMergeResponse>> mergeDocumentPost(
+			@PathVariable(value = "correlationId", required = true) String correlationId, 
+			@Valid @RequestBody(required = true)  DocMergeRequest request)  {
+		
+		logger.info("Starting merge process...");
```
Comment (akroon3r) [inline src/src/main/java/ca/bc/gov/open/pssg/docmerge/controller/MergeController.java:49]: do we want to include correlation id here

File: src/src/main/java/ca/bc/gov/open/pssg/docmerge/controller/MergeController.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+		
+		try {
+			
+			DocMergeResponse mergResp = mergeService.mergePDFDocuments(request, correlationId);
+			JSONResponse<DocMergeResponse> resp = new JSONResponse<>(mergResp);
+			logger.info("Merge process complete.");
```
Comment (akroon3r) [inline src/src/main/java/ca/bc/gov/open/pssg/docmerge/controller/MergeController.java:55]: same

File: src/src/main/java/ca/bc/gov/open/pssg/docmerge/swagger2/SwaggerConfig.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+    private ApiInfo getApiInfo() {
+
+        return new ApiInfoBuilder()
+                .title("Document Merge RESTful Service")
+                .description("Document Merge API")
+                .version(properties.getServiceApiVersion())
+                .build();
+    }
+}
```
Comment (TayGov) [inline src/src/main/java/ca/bc/gov/open/pssg/docmerge/swagger2/SwaggerConfig.java:71]: newline

File: src/src/main/java/ca/bc/gov/open/pssg/docmerge/controller/MergeController.java [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+			return new ResponseEntity<>(resp, HttpStatus.OK);
+			
+		} catch (MergeException e) {
+			
+			e.printStackTrace();
+			logger.error("Document Merge encountered an error " + e.getMessage());
```
Comment (akroon3r) [inline src/src/main/java/ca/bc/gov/open/pssg/docmerge/controller/MergeController.java:61]: same

**Review by TayGov — APPROVED** [review]
Approved but there are things to be addressed

File: src/src/main/java/ca/bc/gov/open/pssg/docmerge/exception/DocMergeExceptionHandler.java [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+		
+	    Map<String, String> errors = new HashMap<>();
+	    e.getBindingResult().getAllErrors().forEach((error) -> {
+	        String fieldName = ((FieldError) error).getField();
+	        String errorMessage = error.getDefaultMessage();
+	        errors.put(fieldName, errorMessage);
+	    });
+	    
+	    StringBuffer buffer = new StringBuffer();
+	    int c = 0;
```
Comment (akroon3r) [inline src/src/main/java/ca/bc/gov/open/pssg/docmerge/exception/DocMergeExceptionHandler.java:81]: I forget, can you define counters in the for loop definition?

File: src/src/main/java/ca/bc/gov/open/pssg/docmerge/exception/MergeException.java [hunk 10]
```diff
@@ -0,0 +1,32 @@
+package ca.bc.gov.open.pssg.docmerge.exception;
+
+import org.springframework.http.HttpStatus;
+
+/**
+ * 
+ * Custom exception for Merge Service. 
+ * 
+ * @author shaunmillargov
```
Comment (akroon3r) [inline src/src/main/java/ca/bc/gov/open/pssg/docmerge/exception/MergeException.java:9]: edit

File: src/src/main/java/ca/bc/gov/open/pssg/docmerge/exception/DocMergeExceptionHandler.java (hunk 9 above)
Comment (TayGov) [reply src/src/main/java/ca/bc/gov/open/pssg/docmerge/exception/DocMergeExceptionHandler.java:81]: yes

File: src/src/main/java/ca/bc/gov/open/pssg/docmerge/model/DocMergeResponse.java (hunk 4 above)
Comment (akroon3r) [inline src/src/main/java/ca/bc/gov/open/pssg/docmerge/model/DocMergeResponse.java:36]:
```suggestion
}

```

File: src/src/main/java/ca/bc/gov/open/pssg/docmerge/model/JSONError.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonPropertyOrder;
+
+/**
+ * 
+ * Error object
+ * 
+ * @author shaunmillargov
```
Comment (akroon3r) [inline src/src/main/java/ca/bc/gov/open/pssg/docmerge/model/JSONError.java:11]: edit

File: src/src/main/java/ca/bc/gov/open/pssg/docmerge/model/JSONResponse.java [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import ca.bc.gov.open.pssg.docmerge.utils.DocMergeConstants;
+
+
+/**
+ * 
+ * JSON Enveloped Response Type (Variable object payload) 
+ * 
+ * @author shaunmillargov
```
Comment (akroon3r) [inline src/src/main/java/ca/bc/gov/open/pssg/docmerge/model/JSONResponse.java:16]: edit

File: src/src/main/java/ca/bc/gov/open/pssg/docmerge/model/MergeDoc.java [hunk 13]
```diff
@@ -0,0 +1,49 @@
+package ca.bc.gov.open.pssg.docmerge.model;
+
+import java.util.UUID;
+
+/**
+ * 
+ * Represents a single input document for merging.
+ * 
+ * @author shaunmillargov
```
Comment (akroon3r) [inline src/src/main/java/ca/bc/gov/open/pssg/docmerge/model/MergeDoc.java:9]: edit

File: src/src/main/java/ca/bc/gov/open/pssg/docmerge/model/Options.java [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+import javax.validation.constraints.Pattern;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonPropertyOrder;
+
+
+/**
+ * Basic option flags for merge process. 
+ * 
+ * @author shaunmillargov
```
Comment (akroon3r) [inline src/src/main/java/ca/bc/gov/open/pssg/docmerge/model/Options.java:13]: edit

File: src/src/main/java/ca/bc/gov/open/pssg/docmerge/service/MergeService.java [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+
+import ca.bc.gov.open.pssg.docmerge.exception.MergeException;
+import ca.bc.gov.open.pssg.docmerge.model.DocMergeRequest;
+import ca.bc.gov.open.pssg.docmerge.model.DocMergeResponse;
+
+/**
+ * 
+ * PDF Merging Service Interface 
+ * 
+ * @author shaunmillargov
```
Comment (akroon3r) [inline src/src/main/java/ca/bc/gov/open/pssg/docmerge/service/MergeService.java:11]: edit

File: src/src/main/java/ca/bc/gov/open/pssg/docmerge/service/MergeServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+import ca.bc.gov.open.pssg.docmerge.model.DocMergeResponse;
+import ca.bc.gov.open.pssg.docmerge.utils.DDXUtils;
+import ca.bc.gov.open.pssg.docmerge.utils.PDFBoxUtilities;
+import ca.bc.gov.open.pssg.docmerge.utils.DocMergeConstants;
+
+/**
+ * 
+ * PDF Merge service. 
+ * 
+ * @author shaunmillargov
```
Comment (akroon3r) [inline src/src/main/java/ca/bc/gov/open/pssg/docmerge/service/MergeServiceImpl.java:44]: edit

File: src/src/main/java/ca/bc/gov/open/pssg/docmerge/swagger2/SwaggerConfig.java (hunk 7 above)
Comment (akroon3r) [inline src/src/main/java/ca/bc/gov/open/pssg/docmerge/swagger2/SwaggerConfig.java:71]:
```suggestion
}

```

File: src/src/main/java/ca/bc/gov/open/pssg/docmerge/utils/Base64Utils.java [hunk 17]
```diff
@@ -0,0 +1,134 @@
...
+ * Base64 encoder/decoder utils based on Apache Commons Codec.
+ * 
+ * @author shaunmillargov
+ *
+ */
+public class Base64Utils {
+
+	public static void main(String args[]) throws Exception {
+
+		//String sourceFile = "C:\\Users\\176899\\github_projects\\DocumentMerge\\test_files\\pdfsample1.pdf";
```
Comment (akroon3r) [inline src/src/main/java/ca/bc/gov/open/pssg/docmerge/utils/Base64Utils.java:22]: do we need this?

File: src/src/main/java/ca/bc/gov/open/pssg/docmerge/utils/DDXUtils.java [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+import org.w3c.dom.Element;
+
+import ca.bc.gov.open.pssg.docmerge.model.MergeDoc;
+
+
+
+/**
+ * A set of DDX utils. 
+ * 
+ * @author shaunmillargov
```
Comment (akroon3r) [inline src/src/main/java/ca/bc/gov/open/pssg/docmerge/utils/DDXUtils.java:28]: edit

File: src/src/main/java/ca/bc/gov/open/pssg/docmerge/utils/DocMergeConstants.java [hunk 19]
```diff
@@ -0,0 +1,45 @@
+package ca.bc.gov.open.pssg.docmerge.utils;
+
+
+/**
+ * 
+ * Digital Forms Constants. 
+ * 
+ * @author shaunmillargov
```
Comment (akroon3r) [inline src/src/main/java/ca/bc/gov/open/pssg/docmerge/utils/DocMergeConstants.java:8]: edit

File: src/src/main/java/ca/bc/gov/open/pssg/docmerge/utils/DocMergeUtils.java [hunk 20]
```diff
@@ -0,0 +1,27 @@
+package ca.bc.gov.open.pssg.docmerge.utils;
+
+import ca.bc.gov.open.pssg.docmerge.model.JSONError;
+import ca.bc.gov.open.pssg.docmerge.model.JSONResponse;
+
+/**
+ * 
+ * @author shaunmillargov
```
Comment (akroon3r) [inline src/src/main/java/ca/bc/gov/open/pssg/docmerge/utils/DocMergeUtils.java:8]: edit

File: src/src/main/java/ca/bc/gov/open/pssg/docmerge/utils/PDFBoxUtilities.java [hunk 21]
```diff
@@ -0,0 +1,109 @@
...
+import org.slf4j.LoggerFactory;
+import org.springframework.http.HttpStatus;
+
+import ca.bc.gov.open.pssg.docmerge.exception.MergeException;
+
+/**
+ * 
+ * PDF Box Utilities 
+ * 
+ * @author shaunmillargov
```
Comment (akroon3r) [inline src/src/main/java/ca/bc/gov/open/pssg/docmerge/utils/PDFBoxUtilities.java:24]: edit

**Review by akroon3r — COMMENTED** [review]
First pass.

File: src/Still_to_be_done.txt (hunk 2 above)
Comment (akroon3r) [reply src/Still_to_be_done.txt:1]: yea remove this

File: src/src/main/java/ca/bc/gov/open/pssg/docmerge/controller/MergeController.java (hunk 8 above)
Comment (TayGov) [reply src/src/main/java/ca/bc/gov/open/pssg/docmerge/controller/MergeController.java:61]: Also we should remove string concatenation here as well.

File: src/src/main/java/ca/bc/gov/open/pssg/docmerge/service/MergeServiceImpl.java [hunk 22]
```diff
@@ -0,0 +1,193 @@
...
+			    }
+			});
+			
+			// For each document, check if XFA and convert to PDF/A if requested via option. 
+			for (ca.bc.gov.open.pssg.docmerge.model.Document doc: request.getDocuments()) {
+				
+				byte[] thisDoc = Base64Utils.decode(doc.getData().getBytes()); 
+				
+				if ( request.getOptions().getForcePDFAOnLoad() && PDFBoxUtilities.isPDFXfa(thisDoc)) {
+					logger.info("forcePDFA is on and document, order " + doc.getOrder() + ", is XFA. Converting to PDF/A...");
```
Comment (TayGov) [inline src/src/main/java/ca/bc/gov/open/pssg/docmerge/service/MergeServiceImpl.java:94]: String concatenation

File: src/src/main/java/ca/bc/gov/open/pssg/docmerge/service/MergeServiceImpl.java [hunk 23]
```diff
@@ -0,0 +1,193 @@
...
+				
+				if ( request.getOptions().getForcePDFAOnLoad() && PDFBoxUtilities.isPDFXfa(thisDoc)) {
+					logger.info("forcePDFA is on and document, order " + doc.getOrder() + ", is XFA. Converting to PDF/A..."); 
+					
+					//call PDF/A transformation 
+					thisDoc = createPDFADocument(thisDoc, sFactory);
+				}
+				
+				pageList.add( new MergeDoc( thisDoc));
+				logger.info("Loaded page " + doc.getOrder());
```
Comment (TayGov) [inline src/src/main/java/ca/bc/gov/open/pssg/docmerge/service/MergeServiceImpl.java:101]: String concatenation

File: src/src/main/java/ca/bc/gov/open/pssg/docmerge/service/MergeServiceImpl.java [hunk 24]
```diff
@@ -0,0 +1,193 @@
...
+				// Get the key name as specified in the DDX document
+				String keyName = (String) e.getKey();
+				if (keyName.equalsIgnoreCase(DocMergeConstants.DDX_OUTPUT_NAME)) {
+					
+					Object o = e.getValue();
+					outDoc = (Document) o;
+
+					// Save the result PDF file (debugging)
+					//File myOutFile = new File("C:\\Users\\176899\\workspaces\\neon\\pdfmerge\\AEMMergeTest\\output\\out.pdf");
+					//outDoc.copyToFile(myOutFile);
```
Comment (TayGov) [inline src/src/main/java/ca/bc/gov/open/pssg/docmerge/service/MergeServiceImpl.java:144]: remove unused code?

File: src/src/main/java/ca/bc/gov/open/pssg/docmerge/service/MergeServiceImpl.java [hunk 25]
```diff
@@ -0,0 +1,193 @@
...
+		
+		// Create a DocConverterServiceClient object
+		DocConverterServiceClient docConverter = new DocConverterServiceClient(factory);
+		Document inDoc = new Document(inputFile);
+	
+		// Create a PDFAConversionOptionSpec object and set tracking information
+		PDFAConversionOptionSpec spec = new PDFAConversionOptionSpec();
+		
+	    // AEM logging level - suggest turning this off unless issues need debugging on AEM side. 
+		//spec.setLogLevel("INFO");
```
Comment (TayGov) [inline src/src/main/java/ca/bc/gov/open/pssg/docmerge/service/MergeServiceImpl.java:182]: Remove unused code?

File: src/src/main/java/ca/bc/gov/open/pssg/docmerge/utils/DDXUtils.java [hunk 26]
```diff
@@ -0,0 +1,168 @@
...
+
+			transformer = transformerFactory.newTransformer();
+			DOMSource source = new DOMSource(ddx);
+			StreamResult result = new StreamResult(new StringWriter());
+			transformer.transform(source, result);
+			return result.getWriter().toString();
+
+		} catch (TransformerConfigurationException e) {
+			e.printStackTrace();
+			logger.error("Error at DDxUtils.DDXDocumentToString: " + e.getMessage());
```
Comment (TayGov) [inline src/src/main/java/ca/bc/gov/open/pssg/docmerge/utils/DDXUtils.java:163]: String concatenation

File: src/src/main/resources/application.properties [hunk 27]
```diff
@@ -0,0 +1,29 @@
+# PDF Merge API Application properties. 
+# Note: Values prefixed with $ are externalized values and must be set via OS Secrets or if running locally 
+# using environmental variables with same name. 
+server.port=8082
```
Comment (TayGov) [inline src/src/main/resources/application.properties:4]: Add env var?

**Review by TayGov — APPROVED** [review]
This is initial state

**Pull request merged**